ifm-5.4/man/ifm.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH IFM "1" "November 2008" "IFM version 5.3" "User Commands"
.SH NAME
IFM \- Interactive Fiction Mapper
.SH SYNOPSIS
.B ifm
[\fIoptions\fR] [\fIfile\fR...]
.SH DESCRIPTION
IFM is a program for keeping track of your progress through an Interactive
Fiction game.  You can record each room you visit and its relation to other
rooms, the initial locations of useful items you find, and the tasks you
need to perform in order to solve the game.

The IFM mapping commands are designed so that you can easily add to the map
as you explore the game.  You type in the rooms you visit and the
directions you move in to reach other rooms, and IFM calculates the
position of each room relative to the others.  A map can consist of several
independent sections, allowing you to divide up the map however you like.

The IFM task commands, if used, allow you to specify the order in which
game-solving tasks must be done.  The IFM program can then calculate and
print different styles of walkthrough for the game.

IFM can write hardcopy maps directly to PostScript, suitable for printing
or viewing.  It can also write maps in Fig format, which can be viewed and
edited using Xfig, or converted into many other formats.

This man page just presents a summary of the command-line options.  Online
documentation on how to use IFM can be found in the
.B SEE ALSO
section.
.SS "Output options:"
.TP
\fB\-m\fR, \fB\-map\fR[=\fIsections\fR]
Select map output
.TP
\fB\-i\fR, \fB\-items\fR
Select item output
.TP
\fB\-t\fR, \fB\-tasks\fR
Select task output
.TP
\fB\-f\fR, \fB\-format\fR <name>
Select output format
.TP
\fB\-o\fR, \fB\-output\fR <file>
Write output to specified file
.SS "Auxiliary options:"
.TP
\fB\-I\fR (dir)
Prepend directory to search path
.TP
\fB\-S\fR, \fB\-style\fR (name)
Push a style onto the style list
.TP
\fB\-s\fR, \fB\-set\fR (var=val)
Set a customization variable
.TP
\fB\-noinit\fR
Don't read personal init file
.TP
\fB\-w\fR, \fB\-nowarn\fR
Don't print warnings
.TP
\fB\-e\fR, \fB\-errors\fR <num>
Max errors before giving up (default: 10)
.SS "Information options:"
.TP
\fB\-show\fR <type>
Show information
.TP
\fB\-v\fR, \fB\-version\fR
Print program version
.TP
\fB\-h\fR, \fB\-help\fR
This help message
.SS "Output formats (may be abbreviated):"
.TP
ps
PostScript
.TP
fig
Fig drawing commands
.TP
text
Nicely\-formatted ASCII text
.TP
rec
Recording of game commands
.TP
dot
Graph of task dependencies
.TP
raw
Raw ASCII text fields
.TP
tk
Tcl/Tk program commands (tkifm)
.SS "Show options (may be abbreviated):"
.TP
maps
Show map sections
.TP
vars
Show defined variables
.TP
path
Show file search path
.SH ENVIRONMENT
.IP IFMPATH
Colon-separated list of extra directories to search for include files.
.SH FILES
.IP \fC$HOME/.ifmrc\fP
User resource file.

.IP \fC$HOME/ifm.ini\fP
Alternate resource filename (Windows-style).
.SH COPYRIGHT
Copyright \(co Glenn Hutchings <zondo42@googlemail.com>
.PP
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2, or (at your option)
any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.SH "SEE ALSO"
.BR xfig (1) ,
.BR fig2dev (1) ,
.BR dot (1)
.IP /usr/local/doc/ifm/index.html
IFM manual


NEWS
===========================
 Version 5.4 (13 Feb 2009)
===========================

* Fixed build problems on various systems.

===========================
 Version 5.3 (14 Nov 2008)
===========================

* IFM is now hosted at Bitbucket_.

* New ``them`` keyword, which refers to the list of items mentioned in the
  current statement.

* New game-solver variables: ``finish_room``, ``finish_item`` and
  ``finish_task``, which can be a comma-separated list of extra rooms,
  items and tasks which get assigned the ``finish`` attribute (i.e., cause
  the solver to exit).

* New task graph variables: ``task_graph_wrap``, which controls
  word-wrapping when printing graph nodes, and ``task_graph_font``, which
  sets the text font.

* The new keywords ``true`` and ``false`` can be used in place of 1 and 0,
  respectively, when setting variables.

* Both versions of the personal initialization file (``.ifmrc`` and
  ``ifm.ini``) are now read, if they exist.

* There's a new Perl script ``ifm2web``, which converts IFM maps to image
  files suitable for the Interweb.

* The GNU Emacs ``ifm-mode.el`` in the ``contrib`` directory has been
  improved: it can display task recording and debugging output, now has
  ``imenu`` and ``flymake`` support, and a minor bug fix.

* The documentation is now much prettier, thanks to Sphinx_ and Pygments_.
  The ``contrib`` directory contains the Python_ module used to
  syntax-highlight IFM code and IF transcripts.

* Since ``asciimapper`` is now in the IF archive, it is no longer part of
  the ``contrib`` directory.

* PostScript maps now conform properly to Adobe DSC_.  This stops
  Ghostscript 8.x barfing on them.

.. _Sphinx: http://sphinx.pocoo.org
.. _Pygments: http://pygments.org
.. _Python: http://www.python.org
.. _DSC: http://www-cdf.fnal.gov/offline/PostScript/AdobePS.html
.. _Bitbucket: http://www.bitbucket.org/zondo/ifm

==========================
 Version 5.2 (7 Sep 2006)
==========================

* There's a new ``-errors`` option, which controls the maximum number of
  errors printed before giving up.

* New program in the ``contrib`` directory: ``asciimapper``.  This is a
  tool to create IFM maps from ASCII art maps, contributed by Elijah
  Griffin.

* The GNU Emacs ``ifm-mode.el`` in the ``contrib`` directory has new
  commands to display maps, items and tasks and various other things.
  Based on an initial version contributed by `Lee Bigelow`_.

* Several unused features which added needless complexity have been
  removed.  These are: preprocessing by ``gpp``, expressions for
  real/integer values, and variable interpolation in strings.

* An obscure bug has been fixed which involved the ``follow`` and ``need``
  attributes combined.  This caused some valid task orderings to be flagged
  as cycles.

* A bug with modification of room attributes has been fixed.

* Don't crash when trying to solve game and there are no rooms.

* A crash when trying to print an error message about cyclic task
  dependencies has been fixed.

.. _Lee Bigelow: ligelowbee@yahoo.com

===========================
 Version 5.1 (30 Nov 2004)
===========================

* The ``tkifm`` program now supports line numbering and syntax
  highlighting, thanks to a contributor who wishes to remain anonymous.

* There is now a man page summarizing the command-line options.

* In PostScript output, up/down and in/out labels are now drawn separately,
  next to their respective rooms.  Also, if a link leaves a room and
  returns to the same room, it's drawn as a small circle.  Thanks to `Dan
  Eble`_.

* Hopefully, the Windows port should work properly now.

* Fixed weird bug with scaled arcs in PostScript output.  More thanks to
  `Dan Eble`_.

* Fixed bug in ``ifm2tex`` with converting strings to TeX format
  (specifically, ones with embedded zeros).

* Fixed crash when using ``it`` in some circumstances.

.. _Dan Eble: mailto:eble@ticalc.org

===========================
 Version 5.0 (23 Jan 2003)
===========================

* New map output: Fig format, suitable for viewing or editing with Xfig, or
  exporting to other formats.

* New utility program: ``ifm2dev``, which pipes IFM output in Fig format
  through ``fig2dev`` repeatedly to write each map section to its own file.

* New utility program: ``ifm2tex``, which converts Fig files to EPS and
  includes them in a LaTeX document together with tables of items and
  tasks.

* New task output: graphviz (``dot``) format, which dumps a task dependency
  graph.

* New ``keep with`` and ``keep until`` syntax for items, allowing items to
  be kept until certain tasks are done or certain other items are dropped.

* New room ``nodrop`` attribute, indicating that no items should be dropped
  there.

* New task and item ``ignore`` attribute, indicating that these things
  should be ignored when finding a solution.

* New task ``do`` attribute, indicating that when a task is done, it also
  does one or more other tasks automatically.

* New option: ``-style``, which sets global styles for all IFM objects.

* New option: ``-set``, which lets you set variables from the command line.
  To support this, the variable-setting syntax has been changed: instead
  of a space separating the driver name from the variable, there's now a
  period.

* New ``-show`` option: ``maps``, which lists all the map sections and
  their sizes.

* The ``-map`` option now accepts an optional list of map sections to
  print.

* There's a bunch of new variables to control things, including the overall
  behaviour of the game solver.

* IFM input is now *preprocessed*, in a similar way to C programs.  As a
  result, the old ``include`` and ``alias`` features have been replaced
  with the preprocessor equivalents.

* There's a new ``require`` keyword, which lets you say that a certain
  version of IFM (or later) is required for a map.

* Changed implicit-link style behaviour: the ``style`` clause on a room

contrib/pyifhilite/ifhilite/ifm.py
"""
IFM syntax highlighting.
"""

from pygments.lexer import RegexLexer
from pygments.token import (Text, Comment, Operator, Keyword, Name,
                            String, Number, Punctuation, Token)

structure = ["room"]

specials = ["title", "map", "require", "style", "endstyle"]

builtins = ["it", "them", "last", "true", "false", "undef"]

keywords = ["all", "after", "any", "before", "cmd", "d", "do", "down",
            "dir", "drop", "endstyle", "except", "exit", "finish", "follow",
            "from", "get", "give", "go", "goto", "hidden", "ignore", "in",
            "item", "join", "keep", "leave", "length", "link", "lose",
            "lost", "map", "need", "none", "nodrop", "nolink", "nopath",
            "note", "oneway", "out", "require", "room", "safe", "score",
            "start", "style", "tag", "task", "title", "to", "u", "up",
            "until", "with"]

directions = ["n", "north", "ne", "northeast", "e", "east",
              "se", "southeast", "s", "south", "sw", "southwest",
              "w", "west", "nw", "northwest"]

obsolete = ["given", "times"]

def wordlist(list):
    return "(" + "|".join(list) + r")\b"

class IFMLexer(RegexLexer):
    """
    IFM code lexer.
    """

    name = 'Interactive Fiction Mapper'
    aliases = ['ifm', 'IFM']
    filenames = ['*.ifm']
    mimetypes = ['text/x-ifm', 'application/x-ifm']

    tokens = {
        'root': [
            (r'\n', Text),
            (r'[^\S\n]+', Text),
            (r'#.*$', Comment),
            (r';', Punctuation),
            (r'\\\n', Text),
            (r'\\', Text),
            (r'=', Operator),
            (r'"[^"]*"', String),
            (r'^.+?=', Token.Token),

            ("^" + wordlist(structure), Name.Class),
            ("^" + wordlist(specials), Name.Entity),

            (wordlist(keywords), Token.Keyword.Reserved),
            (wordlist(builtins), Name.Builtin),
            (wordlist(directions), Name.Variable),
            (wordlist(obsolete), Token.Generic.Error),

            (r'(\d+\.?\d*|\d*\.\d+)([eE][+-]?[0-9]+)?', Number.Float),
            (r'\d+', Number.Integer),
            (r'[a-zA-Z_][a-zA-Z0-9_.]*', Name),
        ],
    }
